cpp/src_prims/sparse/distance/csr_spmv.cuh

```diff
@@ -0,0 +1,354 @@
+      shared_idx += 1 * run_r;
+      value_t right_side = rv * run_r;
+
+      // Apply semiring "sum" & "product" functions locally
+      cur_sum = accum_func(cur_sum, reduce_func(left_side, right_side));
+
+      // finished when all items in chunk have been
+      // processed
+      done = l == -1 && r == -1;
+
+    } else {
+      done = true;
+    }
+  }
+
+  __device__ inline bool isdone() { return done; }
+
+  __device__ inline void write(value_t *out) {
+    for (int i = tid; i < n_rows; i += blockDim.x) {
+      out[row_a * n + row_b] = cur_sum;
+    }
+  }
+
+ private:
+  int tid;
+
+  bool done;
+
+  int shared_size;
+
+  value_idx n_rows;
+
+  value_t p;
+
+  value_idx local_idx;
+  value_idx local_idx_stop;
+  value_idx shared_idx;
+
+  value_t cur_sum;
+
+  value_idx n_entries;
+
+  value_idx m;
+  value_idx n;
+  value_idx start_offset_a;
+  value_idx stop_offset_a;
+
+  value_idx row_a;
+  value_idx row_b;
+
+  value_idx start_offset_b;
+
+  value_idx start_row_b;
+  value_idx stop_row_b;
+
+  value_idx *offsets_a;
+
+  // shared memory
+  value_idx row_count;
+  value_idx *shared_cols;
+  value_t *shared_vals;
+  value_idx *chunk_cols;
+  value_t *chunk_vals;
+};
+
+/**
+ * Optimized for large numbers of rows but small enough numbers of columns
+ * that each thread can process their rows in parallel.
+ * @tparam value_idx index type
+ * @tparam value_t value type
+ * @tparam tpb block size
+ * @tparam reduce_f semiring product() function
+ * @tparam accum_f semiring sum() function
+ * @param[in] indptrA csr column index pointer array for A
+ * @param[in] indicesA csr column indices array for A
+ * @param[in] dataA csr data array for A
+ * @param[in] indptrB csr column index pointer array for B
+ * @param[in] indicesB csr column indices array for B
+ * @param[in] dataB csr data array for B
+ * @param[in] m number of rows in A
+ * @param[in] n number of rows in B
+ * @param[out] out dense output array of size m * n
+ * @param[in] n_blocks_per_row number of blocks of B scheduled per row of A
+ * @param[in] n_rows_per_block number of rows of A scheduled per block of B
+ * @param[in] buffer_size number of nonzeros to store in smem
+ * @param[in] reduce_func semiring product() function
+ * @param[in] accum_func semiring sum() function
+ */
+template <typename value_idx, typename value_t, int tpb, typename reduce_f,
+          typename accum_f>
+__global__ void classic_csr_semiring_spmv_kernel(
+  value_idx *indptrA, value_idx *indicesA, value_t *dataA, value_idx *indptrB,
+  value_idx *indicesB, value_t *dataB, value_idx m, value_idx n, value_t *out,
+  int n_blocks_per_row, int n_rows_per_block, int buffer_size,
+  reduce_f reduce_func, accum_f accum_func) {
+  value_idx out_row = blockIdx.x / n_blocks_per_row;
+  value_idx out_col_start = blockIdx.x % n_blocks_per_row;
+
+  value_idx row_b_start = out_col_start * n_rows_per_block;
+  value_idx tid = threadIdx.x;
+
+  if (out_row > m || row_b_start > n) return;
```

Comment:
In line 261 `semiring.load_a` is called in which `__syncthreads()` calls are executed. Please delay the return statements after the sync calls are finished. See my comment at coo_spmv.cuh line 116 for more details.